src/gui/SampleWaveform.cpp

```diff
@@ -28,64 +28,39 @@ namespace lmms::gui {
 
 void SampleWaveform::visualize(Parameters parameters, QPainter& painter, const QRect& rect)
 {
-	const int x = rect.x();
-	const int height = rect.height();
-	const int width = rect.width();
-	const int centerY = rect.center().y();
+	if (!parameters.buffer || parameters.size == 0) { return; }
 
-	const int halfHeight = height / 2;
+	const auto centerY = rect.center().y();
+	const auto halfHeight = rect.height() / 2;
 
 	const auto color = painter.pen().color();
 	const auto rmsColor = color.lighter(123);
 
-	const float framesPerPixel = std::max(1.0f, static_cast<float>(parameters.size) / width);
+	const auto samplesPerPixel = std::max(1.0f, static_cast<float>(parameters.size) / rect.width());
+	const auto numPixelsToDraw = std::min(static_cast<int>(parameters.size), rect.width());
 
-	constexpr float maxFramesPerPixel = 512.0f;
-	const float resolution = std::max(1.0f, framesPerPixel / maxFramesPerPixel);
-	const float framesPerResolution = framesPerPixel / resolution;
+	const auto compFn = [](const SampleFrame& a, const SampleFrame& b) { return a.average() < b.average(); };
+	const auto squaredSumFn = [](const float acc, const SampleFrame& x) { return acc + x.average() * x.average(); };
 
-	size_t numPixels = std::min(parameters.size, static_cast<size_t>(width));
-	auto min = std::vector<float>(numPixels, 1);
-	auto max = std::vector<float>(numPixels, -1);
-	auto squared = std::vector<float>(numPixels, 0);
-
-	const size_t maxFrames = static_cast<size_t>(numPixels * framesPerPixel);
-
-	auto pixelIndex = std::size_t{0};
-
-	for (auto i = std::size_t{0}; i < maxFrames; i += static_cast<std::size_t>(resolution))
+	for (auto i = 0; i < numPixelsToDraw; i++)
 	{
-		pixelIndex = i / framesPerPixel;
-		const auto frameIndex = !parameters.reversed ? i : maxFrames - i;
+		const auto start = parameters.buffer + static_cast<int>(std::floor(i * samplesPerPixel));
+		const auto end = parameters.buffer + static_cast<int>(std::ceil((i + 1) * samplesPerPixel));
```

Comment:
```suggestion
		const SampleFrame* start = parameters.buffer + static_cast<int>(std::floor(i * samplesPerPixel));
		const SampleFrame* end = parameters.buffer + static_cast<int>(std::ceil((i + 1) * samplesPerPixel));
```
I think replacing `auto` here would make the code easier to read.